### PR TITLE
Ngmx_328 and NGMX_379 - Add sparse type and ndarray integration to ngraph_bridge

### PR DESCRIPTION
**Pull request by ashokei**
## Description ##
Adds sparse storage type awareness and ndarray integration to ngraph_bridge

### Essentials ###
- [ ✅ ] Passed code style checking (`make lint`)
- [ ✅ ] Changes are complete (i.e. I finished coding on this PR)
- [ ✅ ] All changes have test coverage

### Changes ###
- [ ✅ ] Sparse storage type awareness
- [ ✅ ] NDArray interface integration

## Comments ##
- test_gluon.py:test_export unittest fails with minor accuracy error, same as prior to this PR

### CODE REVIEW

**Comment by ashokei** [comment]
@mbrookhart it should only bypass that specific sparse op/node. Everything else should be in ngraph, it should not bypass entire graph. Should be similar to f16 op or node, if there is one in the middle of a graph, or at the front/input.?

**Comment by mbrookhart** [comment]
! Nevermind, I was looking at the gradient, if it's a sparse delta, you won't use nGraph. It's better than I thought!

**Comment by ashokei** [comment]
yes.. we only do it for "head" gradient that is sparse.

File: src/ngraph/ngraph_compiler.cc [hunk 1]
```diff
@@ -111,6 +114,7 @@ void Compiler::Infer(const SimpleBindArg* simplebind) {
   const auto& idx = graph_.indexed_graph();
   shapes_.resize(idx.input_nodes().size(), nnvm::TShape());
   dtypes_.resize(idx.input_nodes().size(), -1);
+  stypes_.resize(idx.input_nodes().size(), -1);
```
Comment (adstraw) [inline src/ngraph/ngraph_compiler.cc]: use enum (default or undefined) to init

File: src/ngraph/ngraph_compiler.cc [hunk 2]
```diff
@@ -92,17 +92,20 @@ void Compiler::Infer(const BindArg* bind) {
...
       dtypes_.push_back(bind->in_args_[arg_top].dtype());
+      stypes_.push_back(bind->in_args_[arg_top].storage_type());
       ++arg_top;
     }
   }
 
   // append default shapes / dtypes so that vector size = graph size
   shapes_.resize(idx.input_nodes().size(), nnvm::TShape());
   dtypes_.resize(idx.input_nodes().size(), -1);
+  stypes_.resize(idx.input_nodes().size(), mxnet::kDefaultStorage);
```
Comment (adstraw) [inline src/ngraph/ngraph_compiler.cc]: I was unsure when I was working on this... should it be Default or should it be Undefined?

File: src/executor/graph_executor.cc [hunk 3]
```diff
@@ -662,7 +671,7 @@ void GraphExecutor::Init(nnvm::Symbol symbol,
   // This function can be called by regular bind
   // operation flow as well.
 #if MXNET_USE_NGRAPH == 1
-  if (multi_context) {
+  if (!multi_context && !grad_sparse) {
```
Comment (adstraw) [inline src/executor/graph_executor.cc]: I'm confused about the logic change (the '!') for multi_context.  Why did this change?

Comment (mbrookhart) [reply src/executor/graph_executor.cc]: Looks like a typo on my part in the original. Not sure why we haven't hit a bug here.

Comment (ashokei) [reply src/executor/graph_executor.cc]: i believe this was missing originally , @mbrookhart  can clarify.

File: src/ngraph/ngraph_compiler.cc (hunk 1 above)
Comment (mbrookhart) [reply src/ngraph/ngraph_compiler.cc]: It's a data member, right? We don't know the size until this function is executed?

File: src/ngraph/ngraph_compiler.cc (hunk 2 above)
Comment (ashokei) [reply src/ngraph/ngraph_compiler.cc]: nothing of sparse-type comes inside ngraph-bridge, we initialize to default value (kDefaultStorage) based on comment above. Undefined should be ok too, as i expect this to get initialized to kdefaultstorage at any point.

File: src/ngraph/ngraph_compiler.cc (hunk 1 above)
Comment (ashokei) [reply src/ngraph/ngraph_compiler.cc]: thanks, will make this change

File: src/ngraph/ngraph_compiler.cc (hunk 2 above)
Comment (adstraw) [reply src/ngraph/ngraph_compiler.cc]: I think either would work but I am fine with default.

**Pull request merged**